### PR TITLE
Updates the optimization tutorial for TensorFlow 2.0

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The existing TensorFlow tutorial was out of date (requiring the old TF1.3 backend), and hand-written using restructured Text.

**Description of the Change:**

* Ports the tutorial to use the new TF2 backend

* Adds a section to the tutorial on computing quantum gradients using the TF2 backend

* Ports the tutorial to Sphinx-Gallery

**Benefits:** As the tutorial has been ported to Sphinx gallery, it is now executable, making it easy to test to see whether it provides the correct output.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
Rendered version of the tutorial [available here](https://user-images.githubusercontent.com/2959003/80346644-35ffdd00-88aa-11ea-99d6-1ee238c4eccc.png).

File: examples_apps/run_tutorial_machine_learning.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+with prog.context as q:
+    Dgate(alpha) | q
+
+# Assign our TensorFlow variables, so that we can
+# refer to them later when differentiating/training.
+a = tf.Variable(0.43)
+
+with tf.GradientTape() as tape:
+    # Here, we pass map our quantum free parameter `alpha`
```
Comment (thisac) [inline examples_apps/run_tutorial_machine_learning.py:133]:
```suggestion
    # Here, we map our quantum free parameter `alpha`

```

Perhaps adding `and pass it to the engine.` to the sentence instead?

File: examples_apps/run_tutorial_machine_learning.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    # to our TensorFlow variable `a`.
+    result = eng.run(prog, args={"alpha": a})
+    state = result.state
+
+    # Note that all processing, even state-based post-processing,
```
Comment (thisac) [inline examples_apps/run_tutorial_machine_learning.py:138]:
```suggestion
    # Note that all processing, including state-based post-processing,

```

File: examples_apps/run_tutorial_machine_learning.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    # must be done within the gradient tape!
+    mean, var = state.mean_photon(0)
+
+# test the gradient of the variance is correct
```
Comment (thisac) [inline examples_apps/run_tutorial_machine_learning.py:142]:
```suggestion
# test that the gradient of the variance is correct

```

File: doc/introduction/tutorials.rst [hunk 4]
```diff
@@ -68,7 +68,6 @@ Algorithms
     /tutorials/tutorial_post_selection
     /tutorials/tutorial_boson_sampling
     /tutorials/tutorial_gaussian_boson_sampling
-    /tutorials/tutorial_machine_learning
```
Comment (antalszava) [inline doc/introduction/tutorials.rst:71]: How come this was listed, yet it does not render with the previous version of docs?

File: examples/optimization.py [hunk 5]
```diff
@@ -14,23 +11,27 @@
...
+    # reset the engine if it has already been executed
+    if eng.run_progs:
+        eng.reset()
 
-# Create Tensorflow Session and initialize variables
-sess = tf.Session()
-sess.run(tf.global_variables_initializer())
+    with tf.GradientTape() as tape:
+        # execute the engine
+        results = eng.run(circuit, args={'alpha': tf_alpha, 'phi': tf_phi})
```
Comment (antalszava) [inline examples/optimization.py:29]: Was `eval=False` removed since that's the default here?

File: examples/optimization.py [hunk 6]
```diff
@@ -14,23 +11,27 @@
...
+        prob = results.state.fock_prob([1])
+        # negative sign to maximize prob
+        loss = -prob
 
-# Carry out optimization
-for step in range(50):
-    prob_val, _ = sess.run([prob, minimize_op])
-    print("Value at step {}: {}".format(step, prob_val))
+    gradients = tape.gradient(loss, [tf_alpha, tf_phi])
+    opt.apply_gradients(zip(gradients, [tf_alpha, tf_phi]))
```
Comment (antalszava) [inline examples/optimization.py:36]: Could [`tf.minimize`](https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L289) be used here for brevity?

Based on the docstrings, it seems that the `tf.GradientTape()` + `apply_gradients` method approach is useful when the `gradients` are being processed (I might be missing something here!).

File: examples_apps/run_tutorial_machine_learning.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+a = tf.Variable(0.43)
+
+with tf.GradientTape() as tape:
+    # Here, we pass map our quantum free parameter `alpha`
+    # to our TensorFlow variable `a`.
+    result = eng.run(prog, args={"alpha": a})
+    state = result.state
+
+    # Note that all processing, even state-based post-processing,
+    # must be done within the gradient tape!
```
Comment (antalszava) [inline examples_apps/run_tutorial_machine_learning.py:139]: (Minor suggestion)
```suggestion
    # must be done within the gradient tape context!
```

File: examples_apps/run_tutorial_machine_learning.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    mean, var = state.mean_photon(0)
+
+# test the gradient of the variance is correct
+grad = tape.gradient(mean, [a])
```
Comment (antalszava) [inline examples_apps/run_tutorial_machine_learning.py:143]: To match the previous comment
```suggestion
grad = tape.gradient(var, [a])
```

File: examples_apps/run_tutorial_machine_learning.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+#
+# Parameters supplied to a circuit in batch-mode operation can either be scalars or vectors
+# (of length ``batch_size``). Scalars are automatically broadcast over the batch dimension.
+
+alpha = tf.Variable([0.5] * batch_size)
+theta = tf.constant(0.0)
+phi = tf.Variable([0.1, 0.33, 0.5])
+
+##############################################################################
+# Measurement results will be returned as Tensors with shape ``(batch_size,)``.
```
Comment (antalszava) [inline examples_apps/run_tutorial_machine_learning.py:225]: (Minor suggestion): or comparable to refer to the specific type of tensor
```suggestion
# Measurement results will be returned as ``tf.Tensor``s with shape ``(batch_size,)``.

```

File: examples_apps/run_tutorial_machine_learning.py [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+# where photons bunch together in the same mode. Your circuit should contain two modes, each with a
+# single-photon input state, and a beamsplitter with variable parameters. By optimizing the beamsplitter
+# parameters, minimize the probability of the :math:`|1,1\rangle` Fock-basis element of the output state.
+#
+# .. rubric:: Footnotes
+#
+# .. [#] Note that certain operations---in particular, measurements---may not have gradients defined within
+#        Tensorflow. When optimizing via gradient descent, we must be careful to define a circuit which is end-to-end differentiable.
+#
+# .. [#] Note that ``batch_size`` should not be set to 1. Instead, use ``batch_size=None``, or just
```
Comment (antalszava) [inline examples_apps/run_tutorial_machine_learning.py:311]: Is this being checked explicitly?

File: examples/optimization.py (hunk 5 above)
Comment (josh146) [reply examples/optimization.py:29]: Removed since it no longer exists 🙂 In TF2, there is no concept of unevaluated tensors, all tensors are evaluated.

File: examples/optimization.py (hunk 6 above)
Comment (josh146) [reply examples/optimization.py:36]: Oh good catch! This didn't exist back when I first started using eager mode 😆

File: examples_apps/run_tutorial_machine_learning.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+with prog.context as q:
+    Dgate(alpha) | q
+
+# Assign our TensorFlow variables, so that we can
+# refer to them later when differentiating/training.
+a = tf.Variable(0.43)
+
+with tf.GradientTape() as tape:
+    # Here, we pass map our quantum free parameter `alpha`
+    # to our TensorFlow variable `a`.
```
Comment (josh146) [inline examples_apps/run_tutorial_machine_learning.py:134]:
```suggestion
    # to our TensorFlow variable `a` and pass it to the engine.

```

File: examples_apps/run_tutorial_machine_learning.py (hunk 8 above)
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:143]: Oops, this should be mean, it was a typo above saying variance!

File: examples_apps/run_tutorial_machine_learning.py [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+print("Exact gradient:", 2*a)
+print("Exact and TensorFlow gradient agree:", np.allclose(grad, 2*a))
+
+##############################################################################
+# Processing data
+# ---------------
+#
+# The parameters for Blackbird states, gates, and measurements may be more complex
+# than just raw data or machine learning weights. These can themselves be the outputs
+# from some learnable function like a neural network [#]_.
```
[the remark this answers is not in the document]
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:160]: Because it was written using TensorFlow 1.0, and I'm not sure it is relevant any more (but thought it was nice to keep the intro text). In TensorFlow 1.0, creating the neural network was more complicated, as you were constructing a symbolic and unevaluated graph using `tf.placeholders`. Now, creating the neural network is no more complicated than simply matrix multiplication, and feeding the output of that to the Dgate.

File: examples_apps/run_tutorial_machine_learning.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+print(np.sin(q0)**2)
+
+##############################################################################
+# Working with batches
+# --------------------
+#
+# It is common in machine learning to process data in *batches*. Strawberry Fields supports both
+# unbatched and batched data when using the Tensorflow backend. Unbatched operation is the default
+# behaviour (shown above). To enable batched operation, you should provide an extra
+# ``batch_size`` argument [#]_ within the ``backend_options`` dictionary, e.g.,
```
[the remark this answers is not in the document]
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:198]:
```suggestion
# ``batch_size`` argument within the ``backend_options`` dictionary [#]_, e.g.,

```

File: examples_apps/run_tutorial_machine_learning.py (hunk 9 above)
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:225]: unfortunately this won't work, as final backticks must always be followed by spaces!

File: examples_apps/run_tutorial_machine_learning.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+        # execute the engine
+        results = eng.run(circuit, args={'alpha': tf_alpha, 'phi': tf_phi})
+        # get the probability of fock state |1>
+        prob = results.state.fock_prob([1])
+        # negative sign to maximize prob
+        loss = -prob
+
+    gradients = tape.gradient(loss, [tf_alpha, tf_phi])
+    opt.apply_gradients(zip(gradients, [tf_alpha, tf_phi]))
+    print("Value at step {}: {}".format(step, prob))
```
[the remark this answers is not in the document]
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:280]: Unfortunately not, as Sphinx gallery tutorials should be fully executable Python scripts, so you can't use restructured text directives for executable parts of the tutorial.

The reason for the previous duplication (having `examples/optimization.py` as well as `doc/tutorials/tutorials_machine_learning.rst` was that the ReST tutorial wasn't executable, and we wanted to have a quick script that users could also execute. However, now that the tutorial _is_ the executable script, the example is likely redundant and can be removed.

File: examples_apps/run_tutorial_machine_learning.py (hunk 10 above)
Comment (josh146) [reply examples_apps/run_tutorial_machine_learning.py:311]: Good question, I'm not sure!

File: doc/requirements.txt [hunk 15]
```diff
@@ -14,7 +14,7 @@ sphinx-copybutton
 sphinx-automodapi
 sphinx-gallery==0.5.0
 sphinxcontrib-bibtex==0.4.2
-tensorflow-tensorboard>=0.1.8
-tensorflow==1.3
+tensorflow-tensorboard==0.1.8
+tensorflow==2.0
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/requirements.txt:18]: Generally, `requirements.txt` should be a strict `==`, as it defines the base development environment that all devs/services should be using. `setup.py`, on the other hand, should avoid pinning wherever possible.

In this case, we don't do a very good job with the top-level `requirements.txt` file --- we should update it to pin everything!

File: examples/optimization.py (hunk 6 above)
Comment (josh146) [reply examples/optimization.py:36]: Have added a section using `tf.minimize` :) Note that we can't fully replace the existing approach, as `tf.minimize` doesn't allow us to extract intermediate values (such as the probability), but it's nice to show both methods.
